README.md
# Travelling Salesman Problem
Implementation of a Genetic Algorithm using Clojure. See the literate programming page: [Docs](https://vascoferreira25.github.io/travelling-salesman-problem/index.html)

> "Given a list of cities and the distances between each pair of cities, what
> is the shortest possible route that visits each city and returns to the
> origin city?"
- [Wikipedia](https://en.wikipedia.org/wiki/Travelling_salesman_problem)

This implementations uses a genetic algorithm to find the shortest route
between cities.

## Installation

1. Clone this repo:
```shell
git clone https://github.com/vascoferreira25/travelling-salesman-problem
```

## Run the program
First make the project folder your working directory:
```shell
cd travelling-salesman-problem
```

And then, either run it directly with:
```shell
lein run
```

or use the REPL and execute the main function:
```shell
lein repl
```

uncomment the last line
```clojure
(-main)
```

and execute the function.

## Docs and Literate Programming

This project has a literate programming page that walks the entire code to explain
everything that is done at each step.

Check it at: [Docs](https://vascoferreira25.github.io/travelling-salesman-problem/index.html)

## Purpose

Find the best possible route for a list of cities.
The cities are represented in the following chart:

![cities_map](./resources/cities_map.png "Map with the cities")

## Results

With the following parameters, the algorithm reached the result of `565.685424949238`:
- Generations: 200,
- Population Size: 500,
- Mutation Rate: 0.02.

### Evolution
![evolution_chart](./resources/evolution_chart.png "Best Distance Evolution per Generation")

### Solution
![solution](./resources/solution.png "Solution found by the GA algorithm.")


## License

Copyright © 2019

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 1.2.0 - 2019-11-13
### Fixed
- The algorithm was not updating the fitness and total distance values after mutation.
- Each individual has a normalized fitness value instead of replacing the fitness value.
- Fixed documentation.

### Changed
- Swap-cities: now it iterates over the cities and swap them with a mutation rate probability.
- Selection: now it uses both random selection and elite selection.

## 1.1.0 - 2019-10-29
### Fixed
- Fixed the documentation.

### Added
- Literate programming documentation page.

## 1.0.0 - 2019-10-28
### Added
- Generic Algorithm to solve the Travelling Salesan Problem.

[Unreleased]: https://github.com/your-name/travelling-salesman-problem/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/travelling-salesman-problem/compare/0.1.0...0.1.1


test/travelling_salesman_problem/core_test.clj
(ns travelling-salesman-problem.core-test
  (:require [clojure.test :refer :all]
            [travelling-salesman-problem.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/travelling_salesman_problem/utils.clj
(ns travelling-salesman-problem.utils
  (:gen-class)
  (:require [criterium.core :as criterium]
            [uncomplicate.neanderthal
    [native :refer [dv ; double precision vector
                    dge ; double precision dense matrix
                    fv ; float vector
                    fge ; float dense matrix
                    native-double ; default double precision factory
                    native-float]] ; default single precision factory
    [core :refer [submatrix ; returns submatrix of matrix a, from
                            ; row i, column, j, that has k rows and l columns
                  native ; ensures that x is the native main memory
                  iamax ; index of the absolute max value
                  imax ; the index of the max value
                  iamin ; index of the absolute min value
                  imin ; the index of the min value
                  asum ; absolute value of all the elements
                  sum ; sum of the elements
                  dot ; dot product of matrices x and y
                  dim ; returns the dimension of vector x
                  mm ; matrix multiplication - returns new instance
                  ge ; dense matrix in the context of factory
                  scal ; multiplies all entries of x by scaler alpha
                  trans ; transposes matrix
                  trans! ; transposes matrix in-place
                  col ; returns the j column of matrix a
                  cols ; returns lazy sez of column vectors of a
                  row ; returns the i row of matrix a
                  rows ; returns lazy seq of row vectors of a
                  ]]
    [math :refer [abs ; absolute
                  sqrt ; square root
                  log ; logarithm
                  sin ; sin function
                  cos ; cos function
                  pi ; pi number
                  pow ; power x by y
                  sqr]]
    [random :refer [rand-normal! ; populate vector with normal distributed values
                                 ; rng-state, mu (average), sigma (std-dev), x
                    rand-uniform! ; populate vector with uniform distributed values
                                  ; rng-state, [lower], [upper], x
                    rng-state]]]))

(defn shuffle-vec
  "Shuffle a vector randomly."
  [v]
  (loop [vec-left v
         shuffled-vec []]
    (if (empty? vec-left)
      shuffled-vec
      (let [vec-count (count vec-left)
            r (rand-int vec-count)
            sub-1 (subvec vec-left 0 r)
            sub-2 (subvec vec-left (inc r))
            vl (vec (concat sub-1 sub-2))]
        (recur vl (conj shuffled-vec (nth vec-left r)))))))


(defn )


src/travelling_salesman_problem/core.clj
;; # Travelling Salesman Problem
;; This is an alternative implementation in Clojure of the
;; Python tutorial in [Evolution of a salesman: A complete genetic algorithm tutorial for Python](https://towardsdatascience.com/evolution-of-a-salesman-a-complete-genetic-algorithm-tutorial-for-python-6fe5d2b3ca35)
;; And also changed a few details as in [Coding Challenge #35.4: Traveling Salesperson with Genetic Algorithm](https://www.youtube.com/watch?v=M3KTWnTrU_c)

;; ## The Problem The travelling Salesman Problem asks que following question:
;; "Given a list of cities and the distances between each pair of cities, what
;; is the shortest possible route that visits each city and returns to the
;; origin city?"
;; - [Wikipedia](https://en.wikipedia.org/wiki/Travelling_salesman_problem)
;;
;; This implementations uses a genetic algorithm to find the shortest route
;; between cities.

;; -----
;; ## Requirements
;; - Neanderthal: math functions
;; - Clojure.java.io: write data to file
;; - Clojure.string: join all the data in a single string
(ns travelling-salesman-problem.core
  (:gen-class)
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [uncomplicate.neanderthal
             [math :refer [abs sqrt pow]]]))


;; -----
;; ## Records
;; Create a record for Cities and Individuals.
;; This is the same as creating a class for each one.
;; In Clojure it is the same as a simple map.

;; The City is a point and has 3 keys, a `:name` and a `:x` and `:y` coordinates.
(defrecord City [name x y])

;; The Individual represents the genomes in the Genetic Algorithm.
;; Each individual has a parent (except on the initial population)
;; and the `:parents` key may be used to track its evolution in each
;; generation.
(defrecord Individual [route total-distance fitness normalized-fitness])


;; -----
;; ## Utility Functions

(defn dir?
  "Check if the given directory path exists."
  [path]
  (.isDirectory (java.io.File. path)))


(defn mkdir
  "Create a directory."
  [path]
  (.mkdirs (io/file path)))


(defn ensure-directory!
  "If the path doesn't exist, create it."
  [path]
  (when-not (dir? path)
    (mkdir path)))


(defn save-historical-data
  "Save the evolution data into a CSV file for statistics purposes.
  Example path: \"./ga_output\"."
  [path file-name data]
  (ensure-directory! path)
  (let [d (str/join "\n" data)]
    (with-open [w (io/writer (str path "/" file-name) :append true)]
      (.write w d))))


;; As the cities are just points, just calculate the distance
;; between two points with `sqrt((xa - xb)^2 + (ya - yb)^2)`.
(defn distance-between-cities
  "Calculate the distance between two cities."
  [city-pair]
  (let [city-a (first city-pair)
        city-b (second city-pair)
        dist-x (abs (- (:x city-a) (:x city-b)))
        dist-y (abs (- (:y city-a) (:y city-b)))]
    (sqrt (+ (pow dist-x 2) (pow dist-y 2)))))

;; Distance between the first and second city:
;;
;;     (distance-between-cities [(first cities) (second cities)])
;;     => 10.04987562112089


(defn total-distance
  "Calculate the total distance of the route."
  [route]
  (let [city-pairs (mapv #(vector %1 %2) (butlast route) (rest route))
        first-city (first route)
        last-city (last route)
        distances (map distance-between-cities city-pairs)
        ;; As city-pairs is a sequence of pairs from the first city to the last,
        ;; it misses the last pair which is last-city -> first-city. To make
        ;; this last calculation, calculate the distance between the last city
        ;; and the first to finish the full circuit and sum it to the total.
        last-city-distance (distance-between-cities [last-city first-city])]
    (+ (reduce + distances) last-city-distance)))

;; Total distance of the cities-list:
;;
;;     (total-distance cities)
;;     => 921.6346703083029


(defn fitness
  "Calculate the fitness based on the route total distance.
  As distance decreases, fitness increases.
  1 is the absolute best value."
  [total-dist]
  (/ 1.0 (+ 1 (pow total-dist 8))))

;; Fitness of the cities-list:
;;
;;     (fitness (total-distance cities))
;;     => 0.0010850286260015397


(defn normalize-fitness
  "Normalize fitness based on the max fitness of the population.
  This makes each fitness value correspond to a probability."
  [individual population]
  (let [fit (:fitness individual)
        pop-fitness (map :fitness population)
        max-fitness (apply max pop-fitness)]
    (assoc individual :normalized-fitness (/ fit max-fitness))))


(defn generate-route
  "Generate a random route."
  [city-list]
  (shuffle city-list))


(defn generate-initial-population
  "Generate the initial population for the genetic algorithm with random routes."
  [population-size cities-list]
  (repeatedly population-size
              (fn []
                (let [route (generate-route cities-list)
                      distance (total-distance route)
                      fit (fitness distance)]
                  (Individual. route distance fit 0)))))


(defn select-parents
  "Select a pair of parents from the population"
  [population]
  (loop [pop population
         parent-pairs []]
    (if (empty? pop)
      parent-pairs
      (recur (drop 2 pop) (conj parent-pairs (take 2 pop))))))


(defn combine-routes
  "Combine the cities not present in the first route
  with the cities of the second route."
  [first-route second-route]
  (let [fr-cities (set (map :name first-route))]
    (loop [sl second-route
           new-route first-route]
      (if (empty? sl)
        (do
          new-route)
        (let [city (first sl)
              is-present (contains? fr-cities (:name city))]
          (if is-present
            (recur (rest sl) new-route)
            (recur (rest sl) (conj new-route city))))))))


;; -----
;; ## Elitism
;; ### Select the best performing individuals.

(defn roulette-selection
  "With a random probability, pick the best individuals more often and
  the worse individuals less often."
  [population population-size]
  (loop [cnt 1
         selected-pop []]
    (if (> cnt population-size)
      selected-pop
      (let [r (rand)
            pop (filter #(> (:normalized-fitness %) r) population)
            individual (rand-nth pop)]
        (recur (inc cnt) (conj selected-pop individual))))))


(defn selection
  "Select the population based on roulette selection."
  [population population-size elitism-size]
  (let [elites (take elitism-size (reverse (sort-by :fitness population)))
        selection (roulette-selection population (- population-size elitism-size))]
    (into elites selection)))


;; -----
;; ## Crossover
;; ### Combine a pair of individuals into a new one

(defn crossover-population
  "Generate a new individual based on 2 parents.
  Select a random number of genes of parent 1 and fill with genes from parent 2."
  [population population-size]
  (for [i (range population-size)]
    (let [parent-1 (rand-nth population)
          parent-2 (rand-nth population)
          route-1 (take (rand-int (count (:route parent-1))) (:route parent-1))
          route-2 (:route parent-2)
          new-route (combine-routes route-1 route-2)
          distance (total-distance new-route)
          fit (fitness distance)]
      (Individual. new-route distance fit 0))))


;; -----
;; ## Mutation
;; ### Change de genes of an individual.

(defn swap-cities
  "Swap the order of some elements randomly for an individual given a probability."
  [individual mutation-rate]
  (loop [swap-counter 0
         new-route (vec (:route individual))]
    (if (> swap-counter (count new-route))
      ;; If already looped over all the cities in the route
      ;; stop the loop
      (assoc individual :route new-route)
      (let [r (rand)]
        (if (< r mutation-rate)
          ;; swap cities within the probability of mutation rate
          (let [c-1 (rand-int (count new-route))
                c-2 (rand-int (count new-route))
                city-1 (nth new-route c-1)
                city-2 (nth new-route c-2)
                swap-cities (vec (assoc new-route c-1 city-2 c-2 city-1))]
            (recur (inc swap-counter) swap-cities))
          ;; if not swapped recur the unchanged route
          (recur (inc swap-counter) new-route))))))


(defn swap-and-update
  "Swap the cities of each individual and update its distance and fitness values."
  [individual mutation-rate]
  (let [new-individual (swap-cities individual mutation-rate)
        new-distance (total-distance (:route new-individual))
        new-fitness (fitness new-distance)]
    (assoc new-individual :total-distance new-distance :fitness new-fitness)))


(defn mutate-population
  "Mutate each individual of a population."
  [population cities-list mutation-rate]
  (map #(swap-and-update % mutation-rate) population))


(defn new-generation
  "Generate a new generation of a given population.
  A new generation is created based on selected individuals, which are crossed
  in pairs and then mutated."
  [population cities-list global-best population-size elitism-size mutation-rate]
  (let [pop (conj population global-best) ; with this, new generations will
                                          ; not forget the best individual
        normalized-population (map #(normalize-fitness %1 pop) pop)
        selected-population (selection normalized-population population-size elitism-size)
        crossed-population (crossover-population selected-population population-size)]
    (mutate-population crossed-population cities-list mutation-rate)))


;; -----
;; ## Genetic Algorithm
;;
;; The algorithms works as follows:
;; - Start with a random initial-population
;; - Normalize the fitness values
;; - Start a new generation
;; - Select the best individuals
;; - Crossover the individuals
;; - Mutate them
;; - Update values
;; - Loop

(defn genetic-algorithm
  "Run the genetic algorithm."
  [initial-population
   cities-list
   generations
   population-size
   elitism-size
   mutation-rate
   print-progress]
  (loop [generation 1
         population initial-population
         historical-best-distance []
         historical-distance []
         historical-fitness []
         global-best (first population)]
    (let [current-generation (map #(normalize-fitness %1 population) population)
          ;; Dont use normalized fitness values to select the best performing
          ;; individual as it is a percentage of the current population
          ;; and does not reflect the global best value.
          sorted-gen (sort-by :fitness population)
          gen-best (last sorted-gen)]
      (if print-progress
        (println (str "Generation: " generation
                      ", Best distance: " (:total-distance global-best)
                      ", Best fitness: " (:fitness global-best))))
      (if (= generation generations)
        {:population current-generation
         :historical-best-distance historical-best-distance
         :historical-distance historical-distance
         :historical-fitness historical-fitness
         :global-best global-best}
        (recur
         (inc generation)
         (new-generation current-generation
                         cities-list
                         global-best
                         population-size
                         elitism-size
                         mutation-rate)
         (conj historical-best-distance (:total-distance global-best))
         (conj historical-distance (:total-distance gen-best))
         (conj historical-fitness (:fitness gen-best))
         (if (> (:fitness gen-best) (:fitness global-best))
           gen-best
           global-best))))))


;; -----
;; ## Variables

(def generations
  "Number of iterations." 200)

(def population-size
  "Number of individuals in each generation" 500)

(def elitism-size
  "Number of best individuals to pass each generation" 100)

(def mutation-rate
  "Rate at which the cities in a route will be swapped" 0.02)

(def cities
  "All the cities to find the route for."
  [(City. "1" -100	0)
   (City. "2" 100	0)
   (City. "3" 0	100)
   (City. "4" 0	-100)
   (City. "5" -90	10)
   (City. "6" -80	20)
   (City. "7" -70	30)
   (City. "8" -60	40)
   (City. "9" -50	50)
   (City. "10" -40	60)
   (City. "11" -30	70)
   (City. "12" -20	80)
   (City. "13" -10	90)
   (City. "14" 10	90)
   (City. "15" 20	80)
   (City. "16" 30	70)
   (City. "17" 40	60)
   (City. "18" 50	50)
   (City. "19" 60	40)
   (City. "20" 70	30)
   (City. "21" 80	20)
   (City. "22" 90	10)
   (City. "23" 90	-10)
   (City. "24" 80	-20)
   (City. "25" 70	-30)
   (City. "26" 60	-40)
   (City. "27" 50	-50)
   (City. "28" 40	-60)
   (City. "29" 30	-70)
   (City. "30" 20	-80)
   (City. "31" 10	-90)
   (City. "32" -10	-90)
   (City. "33" -20	-80)
   (City. "34" -30	-70)
   (City. "35" -40	-60)
   (City. "36" -50	-50)
   (City. "37" -60	-40)
   (City. "38" -70	-30)
   (City. "39" -80	-20)
   (City. "40" -90	-10)])


;; -----
;; ## Main Function

(defn -main
  "Run the genetic algorithm, print the best results and save the evolution-data."
  [& args]
  (println "\nStart Genetic Algorithm")
  (let [initial-population (generate-initial-population population-size cities)
        ga (genetic-algorithm initial-population
                               cities
                               generations
                               population-size
                               elitism-size
                               mutation-rate
                               true)
        best (:global-best ga)]
    (save-historical-data "./ga_output" "historical-best-distance.csv" (:historical-best-distance ga))
    (save-historical-data "./ga_output" "historical-distance.csv" (:historical-distance ga))
    (save-historical-data "./ga_output" "historical-fitness.csv" (:historical-fitness ga))
    (println "\nResults:")
    (println "- Shortest Distance: " (:total-distance best))
    (println "- Best Route: " (map :name (:route best)))
    (println "- Route size" (count (map :name (:route best))))
    (println "- Best Fitness: " (:fitness best))))

;; Execute code in REPL:
;;
;;     (-main)

